PYIC-7811: Reprove identity with F2F

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Reprove identity with F2F

### Why did it change

This change allows a user to use the F2F route to reprove their identity after receiving an intervention.

We store the fact that a F2F journey was started after an intervention on the CriResponse item, and then clear that flag once they've completed the journey.

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-7811](https://govukverify.atlassian.net/browse/PYIC-7811)


[PYIC-7811]: https://govukverify.atlassian.net/browse/PYIC-7811?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-7811: Reprove identity with F2F

This change allows a user to use the F2F route to reprove their identity
after receiving an intervention.

We store the fact that a F2F journey was started after an intervention
on the CriResponse item, and then clear that flag once they've completed
the journey.

Wynndow commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -929,4 +934,20 @@ private void sendProfileMatchedAuditEvent(
+    }
+
+    private void removeReproveIdentityFlag(boolean isF2FComplete, CriResponseItem f2fResponseItem) {
+        if (isF2FComplete && f2fResponseItem.isReproveIdentity()) {
```
Rather than removing the flag like this, I did consider adding in a new state to the journey map to reset the session, including the CRI response, like we do elsewhere. This proved difficult though due to requiring the VCs to still be in the session store for a reuse journey.
If we wanted to do this we could add a new reset type that purely resets the CRI response. This approach also adds more complexity to the journey maps.
This way seemed easier and faster - but happy to be corrected.

Joe-Edwards-GDS replied:
What's the scenario you're trying to avoid by doing this? I guess the inverse risk is that if something blows up afterwards (before the intervention is cleared) then they'll have to start again.

If we add a check for `vcs.isPendingEvcsIdentity()` when skipping, will that help avoid the issue? I.e. so that it only skips if the identity is still pending.

Wynndow replied:
Do you mean the `isF2FComplete` flag. I feel like I did have a good reason before the weekend, but now I can't for the life of me remember what it was...

As this change checks that `isReproveIdentity` flag on the response item, I'm not sure if also checking `isPendingIdentity` is needed. Although it's made we wonder if we could just check that rather than updating the response item...

Wynndow replied:
No - because we want to know if the journey was _started_ in response to an intervention - not if the user happens to come back to core with an pre-existing pending response and a new intervention.

Joe-Edwards-GDS replied:
Yeah exactly :)

I think maybe the scenario that you're worried about is like:
- User gets account intervention
- User reproves identity using F2F
- User's intervention is cleared
- User gets _another_ intervention
- The pending record from the first reprove is still hanging around (with `isReproveIdentity=true`), and the intervention gets immediately cleared

I don't think it's a massive risk tbh, but nice to solve if we can.

My theory with adding `isPendingEvcsIdentity` is that it'll mean you only get to complete the intervention if you're _actually_ returning from F2F (not that you just happen to have a response record kicking around).

Wynndow replied:
yeah, i think that was what i was worried about. But you're right, adding the pending check should solve it. I'll fix it up.

Joe-Edwards-GDS commented on api-tests/features/p2-reprove-identity.feature:
```diff
@@ -11,7 +11,30 @@ Feature: Reprove Identity Journey
+        Then I get an OAuth response
+        When I use the OAuth response to get my identity
+        Then I get a 'P2' identity
+        And an 'IPV_IDENTITY_ISSUED' audit event was recorded [local only]
```
Perhaps ahead of the game, but the audit events that really matter here are `IPV_ACCOUNT_INTERVENTION_START` and `IPV_ACCOUNT_INTERVENTION_END`

Wynndow replied:
Do they exist already?

Joe-Edwards-GDS replied:
Yep!

Joe-Edwards-GDS commented on api-tests/features/p2-reprove-identity.feature:
```diff
@@ -1,7 +1,7 @@
 Feature: Reprove Identity Journey
 
-    Scenario: User needs to reprove their identity
+    Scenario: User reproves with F2F
```
Do you think there's value in having a non-F2F example as well? I'm not sure it's that useful, but it feels a bit strange to have F2F only 🤔 

Now we've added COI, I wonder if we should also have some tests to cover a failing COI check too?

Wynndow replied:
There's an API test on the as yet to be raised PR for PYIC-7810 which should address the COI checks.

We could add another test for a non-f2f journey. It won't test loads more, but it is another path (and probably the more commonly used one).

Joe-Edwards-GDS commented on api-tests/features/p2-reprove-identity.feature:
```diff
@@ -11,7 +11,30 @@ Feature: Reprove Identity Journey
+        Then I get a 'page-face-to-face-handoff' page response
 
+        # Return journey after popping out to the Post Office
+        When I start a new 'medium-confidence' journey and return to a 'page-ipv-reuse' page response
```
Does this need to specify 'with reprove identity'? Otherwise it's as if the intervention was lifted

Wynndow replied:
Yup

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/persistence/item/CriResponseItem.java:
```diff
@@ -27,6 +27,7 @@ public class CriResponseItem implements PersistenceItem {
     private String status;
     private long ttl;
     private List<String> featureSet;
+    private boolean reproveIdentity;
```
What happens for existing records - will this just default to `false`?

Wynndow replied:
Yeah - if there's no value in DDB, it'll come back as `false`

Wynndow committed:
PYIC-7811: Extend API tests

This adds another scenario for a non f2f reproving of identity, and
extends a step to ensure that a f2f joruney returns to core with the
reprove flag still set.

Wynndow committed:
PYIC-7811: Check if identity is pending before skipping

This adds a check to make sure that the identity fetched from EVCS is
pending, before allowing the skip of starting a reprove journey. This is
so that if a user has a lingering CriResponse for some reason, they
won't get to skip the intevention by mistake.

Joe-Edwards-GDS commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -599,4 +605,23 @@ private void sendProfileMatchedAuditEvent(
+        return f2fRequest != null && f2fRequest.isReproveIdentity() && vcBundle.isPendingIdentity();
+    }
+
+    private void removeReproveIdentityFlag(CriResponseItem f2fResponseItem) {
```
Do we still need to do this if we're enforcing that it's only skipped on the first return after a F2F reprove?

Wynndow replied:
gah. I'm not sure why my brain is struggling with this so much...

I suppose we only care if the user manages to achieve a profile - as we wouln't want any further interventions to be skipped. And if they *do* achieve a profile, then they will no longer have a pending identity. So won't skip...

And if they *don't* achieve a profile, then they'll be on an identity proving journey anyway.

So no?

Wynndow committed:
PYIC-7811: Don't clear reprove flag from f2f response

As we're now only allowing a skip of starting a reprove identity journey
if the user has a pending identity in EVCS, we don't need to clear the
flag.

We only care if the user manages to achieve a profile - as we wouln't
want any further interventions to be skipped. And if they do achieve a
profile, then they will no longer have a pending identity. So won't
skip...

And if they don't achieve a profile, then they'll be on an identity
proving journey anyway.

Joe-Edwards-GDS reviewed (approved):
This is super confusing tbh, we'll need to see how fraud use it in practice anyway
